nfexpire.1
.TH nfexpire 1 2006-05-20 "" ""
.SH NAME
nfexpire \- data expiry program
.SH SYNOPSIS
.HP 5
.B nfexpire [options]
.SH DESCRIPTION
.B nfexpire
is used to manage the expiration of old netflow data files, created 
by nfcapd(1) or other data collectors such as sfcapd(1). Data expiration 
is done either by nfcapd(1) in auto expiry mode, or by nfexpire which 
can by run at any time or any desired time interval by cron. nfexpire 
can also be savely run while nfcapd auto expires files, for cleaning up 
full disks etc. nfexpire is sub directory hierarchy aware, and handles 
any format automatically.  For a fast and efficient expiration, nfexpire 
creates and maintains a stat file named \fB.nfstat\fR in the data directory. 
Any \fIdirectory\fR supplied with the options below corresponds to the 
data directory supplied to nfcapd(1) using option -l.

.SH OPTIONS
.TP 3
.B -l \fIdirectory
List current data statistics in directory \fIdatadir\fR.
.TP 3
.B -r \fIdirectory
Rescan the specified directory to update the statfile. To be used only
when explicit update is required. Usually nfexpire takes care itself about
rescanning, when needed.
.TP 3
.B -e \fIdatadir
Expire files in the specified \fIdirectory\fR. Expire limits are taken from
statfile ( see -u ) or from supplied options -s -t and -w. Command line options
overwrite stat file values, however the statfile limits are not changed.
.TP 3
.B -s \fImaxsize
Set size limit for the directory. The specified \fIlimit\fR accepts values such
as 100M, 100MB 1G 1.5G etc. Accpeted size factors are K, KB, M, MB, G, GB and T, TB. 
If no factor is supplied bytes (B) is assumed. A value of 0 disables the max size limit.
.TP 3
.B -t \fImaxlife_time
Sets the max life time for files in the directory. The supplied \fImaxlife_time\fR 
accepts values such as 31d, 240H 1.5d etc. Accpeted time scales are w (weeks) d (days)
H (hours). A value of 0 disables the max lifetime limit. If no scale is given, H (hours)
are assumed.
.TP 3
.B -u \fIdatadir
Updates the max size and lifetime limits, specified by -s -t and -w and stores them
in the statfile as default values. A running nfcapd(1) processs doing auto expiry
will take these new values starting with the next expiry cycle. Running nfexpire
next time doing file expiration will take these new limits unless -s -t or -w are
specified.
.TP 3
.B -w \fIwatermark
Set the water mark in % for expiring data. If a limit is hit, files get expired 
down to this level in % of that limit. If not set, the default is 95%.
.TP 3
.B -h
Print help text on stdout with all options and exit.
.TP 3
.B -p
Directories specified by -e, -l and -r are interpreted as profile directories. Only NfSen will need this option.
.TP 3
.B -Y
Print result in parseable format. Only NfSen will need this option. 
.SH "RETURN VALUE"
Returns 
.PD 0
.RS 4 
0   No error. \fn
.P
255 Initialization failed.
.P
250 Internal error.
.RE
.PD
.SH NOTES
There are two ways to expire files: nfcapd in auto-expire mode ( option -e ) and 
nfexpire running by hand or periodically as cron job. Both ways synchronize access 
to the files, therefore both ways can be run in parallel if required.
.TP 0
.B Expiring by nfcapd in auto-expire mode: option -e
If nfcapd is started with option -e, the auto-expire mode is enabled. After each 
cycle ( typically 5min ) nfcapd expires files according to the limits set with nfexpire 
using options -u -s -t and -w. If initially no limits are set, no files get expired.
.TP 0
.B Expiring by nfexpire
nfexpire can be run at any time to expire files. It automatically syncs up with the 
files created by nfcapd in the mean time since the last expire run, if a nfcapd 
collector process is running for that directory in question and expires the files 
according the limits set.
.TP 0
.B Limits
Files are expired according to two limits: maximum disk space used by all files 
in the directory and maximum lifetime of data files, whatever limit is reached first. 
If one of the limit is hit the expire process will delete files down to the watermark 
of that limit.
.P
.SH "SEE ALSO"
nfcapd(1)
.SH BUGS



NEWS
2010-03-05 v1.6.1
- Fix bug in man page for -t
- Test sampler infos before using them ( nfcapd startup )
- Add sampling tags #34, #35 used by JunOS
- nfexpire: Fix empty .nfsat, when setting limits on an empty directory
- Fix coredump for -B -m (-w) combination
- Optimise some extension map code

2009-12-28 stable v1.6
- Few bug fixes in release candidates rc1, rc2 and rc3

2009-11-16 snapshot-1.6b-20091116
- Update sflow collector with new tags
- Add router IP extension
- Add router ID (engine type/ID) extension

2009-09-30 snapshot-1.6b-20090930
- snapshot bugfix release

2009-11-0801 snapshot-1.6b-20090806
- Add srcmask and dstmask aggregation
- Add csv output mode. -o csv
- Fix some bugs of previous beta
- Add bidirectional aggregation of flows ( -b, -B )
- Add possibility to save aggregated flows into file ( -w )
  Note: This results in a behaviour change for -w in combination
  with aggragation )
- Extend -N ( do not scale numbers ) to all text output not just summary
- Make extension handling more robust for some moody IOSes.
- Remove header lines of -s stat, when using -q ( quiet ) 
  Note: This results in a behaviour change for -N
- Remove -S option from nfdump ( legacy 1.4 compatibility )
- Make use of log (syslog) functions for nfprofile.
- Move log functions to util.c

2009-06-19 snapshot-1.6b-20090717
- Flow-tools converter updated - supports more common elements.
- Sflow collector updated. Supports more common elements.
- Add sampling to nfdump. Sampling is automatically recognised
  in v5 undocumented header fields and in v9 option templates.
  see nfcapd.1(1)
- Add @include option for filter to include more filter files.
- Add flexible aggregation comparable to Flexible Netflow (FNF)
- All new tags can be selected in -o fmt:... see nfdump(1)
- topN stat for all new tags is implemented
- Integrate developer code to read from pcap files into stable
- Update filter syntax for new tags
- Added more v9 tags for netflow v9.
  The detailed tags are listed in nfcapd(1)
  Adding new tags also extended the binary file format with
  data block format 2, which is extension based. File format 
  for version <= 1.5.* ( Data block format 1 ) is read 
  transparently. Data block 2 are skipped by nfdump 1.5.7.
  32bit but AS and interface numbers are supported.
- Add flexible storage option for nfcapd. To save disk space, the 
  data extensions to be stored in the data file are user selectable.
- Added option for multiple netflow stream to same port.
  -n <Ident,IP,base_directory>
  Example: -n router1,192.168.100.1,/var/nfdump/router1
  So multiple -n options may be given at the command line
  Old style syntax still works for compatibility, ( -I .. -l ... )
  but then only one source is supported.
- Move to automake for building nfdump
- Switch scaling factor ( k, M, G ) from 1024 to 1000.
- Make nfdump fully 64bit compliant. ( 8bit data alignments and access )

2009-04-17 stable 1.5.8
- Fix daylight summer time bug, when guessing sub dirs. file access ( -M, -r )
- Bug fixes for 64bits CPUs

2008-02-22 stable-1,5.7
- Add icmp type/code decoding
- Add proper icmp v9 decoding
- Fix memory leaks in -e auto expire mode in nfcapd.
- Fix somee potential dead locks with file locking, when expiring
- Fix multicast bug in nfreplay
- Add hostname lookup for IP addresses in filter.

2007-10-15	stable-1.5.6
- Fix odd CISCO behaviour for ICMP type/code in src port.
- Add fast LZO1X-1 compression option (-z) for output file.
- Add lists for port in syntax -> port in [ 135 137 445]
- Add lists for AS syntax -> as in [ 1024 1025 ]
- Bug fix in filter for syntax 'src as and dst as' 

2007-08-24  stable-1.5.5
- Fix nfprofile bug, nfprofile crashes when last opts line is not valid for 
  some reason.
- Fix potential hand for nfexpire, on empty flow directories.

2007-08-08  snapshot-20070808
- Idents may contain '-' in name.
- Fixed install bugs in Makefile.in and configure.in
- Installs now cleanly on Solaris
- Handle 4byte interface numbers in v9. Quick fix: 4bytes reduced to 2bytes.
- Fix aggregation bug in statistics.
- ftok(3) C library call replaced by more reliable own implementation. 
  Did result in error messages like "Another collector is already running" 
- Fix minor bugs iin file range selction -R.
- Add recursive behaviour for -R <directory>
- New option -i can canche Ident descriptor in data files.

2007-03-12  snapshot-20070312
- Bug fix release of 20070306

2007-03-06  snapshot-20070306
- Fix bug in flist.c. Resulted in a coredump when using sub dirs and -R . ( all files )
- Fix minor bug in nfcapd.c. 
- Extend nfprofile for alerting system of nfsen - special version of profiles
- Extend nfprofile for shadow profiles.

2007-08-10  snapshot-20070110
- Fix some compiler warnings, when compiled on a 64bit LINUX
- Fixes an sflow bug: IP address was printed in wrong direction. ( lower bits first )
- Add new IP addr taging option -T for easy parsing for nfsen lookups
- Add new IP list for massive address filtering:
  syntax: ip in [ 12345 23456 3456 ....]
- Change nfprofile for channel based profiling. This breaks with old nfprofile 
  functionality.
- Remove space from ICMP type/code when followed by an IP address

2006-07-21	snapshot-20060809
- Make nfexpire ready for profile expiration
- Fix bug in nfrpofile. sub dir hierarchy not handled correctly.

2006-07-21	snapshot-20060721
- Add -N option for plain number output in summary line

2006-07-21	snapshot-20060721
- Do recursive file selection when a directory is given by -R

2006-06-14	snapshot-20060621
- Add srcas/dstas/proto aggregation.
  Note: This changes the default aggregation behaviour, but gives more flexibility
- Add tos to element statistics list

2006-06-14	snapshot-20060614
- Add additional stat line at the end of output
- Add new binary nfexpire. Manages data expiry on time and/or size based limits
  Includes new bookkeeping records in nfcapd.  See nfexpire(1)
- Add ICMP type/code decoding in flow listing instead of dst port

nfexpire.c
/*
 *  This file is part of the nfdump project.
 *
 *  Copyright (c) 2004, SWITCH - Teleinformatikdienste fuer Lehre und Forschung
 *  All rights reserved.
 *  
 *  Redistribution and use in source and binary forms, with or without 
 *  modification, are permitted provided that the following conditions are met:
 *  
 *   * Redistributions of source code must retain the above copyright notice, 
 *     this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright notice, 
 *     this list of conditions and the following disclaimer in the documentation 
 *     and/or other materials provided with the distribution.
 *   * Neither the name of SWITCH nor the names of its contributors may be 
 *     used to endorse or promote products derived from this software without 
 *     specific prior written permission.
 *  
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
 *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
 *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
 *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
 *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
 *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
 *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
 *  POSSIBILITY OF SUCH DAMAGE.
 *  
 *  $Author: haag $
 *
 *  $Id: nfexpire.c 9 2009-05-07 08:59:31Z haag $
 *
 *  $LastChangedRevision: 9 $
 *  
 */

#include "config.h"

#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/uio.h>
#include <unistd.h>
#include <dirent.h>
#include <sys/param.h>
#include <fcntl.h>
#include <errno.h>
#include <ctype.h>
#include <time.h>

#ifdef HAVE_FTS_H
#   include <fts.h>
#else
#   include "fts_compat.h"
#define fts_children fts_children_compat
#define fts_close fts_close_compat
#define fts_open  fts_open_compat
#define fts_read  fts_read_compat
#define fts_set   fts_set_compat
#endif

#ifdef HAVE_STDINT_H
#include <stdint.h>
#endif

#include "bookkeeper.h"
#include "nfstatfile.h"
#include "expire.h"
#include "util.h"

static void usage(char *name);

void LogError(char *format, ...);

void CheckDataDir( char *datadir);

channel_t *GetChannelList(char *datadir, int is_profile, int do_rescan);

static void usage(char *name) {
		printf("usage %s [options] \n"
					"-h\t\tThis text\n"
					"-l datadir\tList stat from directory\n"
					"-e datadir\tExpire data in directory\n"
					"-r datadir\tRescan data directory\n"
					"-u datadir\tUpdate expire params from collector logging at <datadir>\n"
					"-s size\t\tmax size: scales b bytes, k kilo, m mega, g giga t tera\n"
					"-t lifetime\tmaximum life time of data: scales: w week, d day, H hour, M minute\n"
					"-w watermark\tlow water mark in %% for expire.\n"
					, name);

} // End of usage

/* 
 * expire.c is needed for daemon code as well as normal stdio code 
 * therefore a generic LogError is defined, which maps in this case
 * to stderr
 */
void LogError(char *format, ...) {
va_list var_args;

	va_start(var_args, format);
	vfprintf(stderr, format, var_args);
	va_end(var_args);

} // End of LogError

void CheckDataDir( char *datadir) {
	if ( datadir ) {
		fprintf(stderr, "Specify only one option out of -l -e -r -u or -p \n");
		exit(250);
	}
} // End of CheckDataDir

static char *AbsolutePath(char *dirname) {
char *path;

	if ( !dirname )
		return NULL;

	if ( dirname[0] == '/' ) // it's already absolute path
		return dirname;

	path = (char *)malloc(MAXPATHLEN);
	if ( !path ) {
		LogError("malloc() error in %s line %d: %s\n", __FILE__, __LINE__, strerror(errno) );
		return NULL;
	}
	memset((void *)path, 0, MAXPATHLEN);
	getcwd(path, MAXPATHLEN-strlen(dirname)-2);	// -2: one for '/' and one for '\0'
	strncat(path, "/", 1);
	strncat(path, dirname, strlen(dirname));
	path[MAXPATHLEN-1] = '\0';
	
	return path;

} // End of AbsolutePath

channel_t *GetChannelList(char *datadir, int is_profile, int do_rescan) {
channel_t **c, *channel;
stringlist_t dirlist;
struct stat stat_buf;
int i;

	// Generate list of directories 
	InitStringlist(&dirlist, 32);
	if ( is_profile ) {
		DIR *PDIR = opendir(datadir);
		struct dirent *entry;
		if ( !PDIR ) {
			fprintf(stderr, "Can't read profiledir '%s': %s\n",datadir, strerror(errno) );
			return NULL;
		}
		while ( ( entry = readdir(PDIR)) != NULL ) {
			char	stringbuf[MAXPATHLEN];
			snprintf(stringbuf, MAXPATHLEN-1, "%s/%s", datadir, entry->d_name);
			stringbuf[MAXPATHLEN-1] = '\0';

			if ( stat(stringbuf, &stat_buf) ) {
				fprintf(stderr, "Can't stat '%s': %s\n",stringbuf, strerror(errno) );
				continue;
			}
			if ( !S_ISDIR(stat_buf.st_mode) ) 
				continue;
	
			// skip all '.' entries -> make .anything invisible to nfprofile
			if ( entry->d_name[0] == '.' )
				continue;
	
			InsertString(&dirlist, stringbuf);
		}
		closedir(PDIR);
	} else {
		InsertString(&dirlist, datadir);
	}

	channel = NULL;
	c = &channel;
	for ( i=0; i<dirlist.num_strings; i++ ) {
		int ret;

		*c = (channel_t *)malloc(sizeof(channel_t));
		if ( !*c ) {
			LogError("malloc() error in %s line %d: %s\n", __FILE__, __LINE__, strerror(errno) );
			return NULL;
		}
		memset((void *)*c, 0, sizeof(channel_t));
		(*c)->next = NULL;
		(*c)->datadir = dirlist.list[i];
		(*c)->do_rescan = do_rescan;

		ret = ReadStatInfo((*c)->datadir, &(*c)->dirstat, CRETAE_AND_LOCK);
		switch (ret) {
			case FORCE_REBUILD:
				printf("Force rebuild requested by stat record in %s\n", (*c)->datadir);
				(*c)->do_rescan = 1;		// file corrupt - rescan
				break;
			case STATFILE_OK:
				break;
			case ERR_NOSTATFILE:	// first rescan bevore expire, if no file exists
				if ( do_rescan == 0 ) {
					printf("Force rebuild to create stat record in %s\n", (*c)->datadir);
					(*c)->do_rescan = 1;
				} // else do_rescan is already set - do not report
				break;
			default:
				exit(250);
		}

		(*c)->books_stat = AccessBookkeeper(&((*c)->books), (*c)->datadir);
		if ( (*c)->books_stat == ERR_FAILED ) {
			fprintf(stderr, "Failed to access bookkeeping record.\n");
			exit(250);
		}

		c = &(*c)->next;
	}

	return channel;

} // End of GetChannelList

int main( int argc, char **argv ) {
struct stat fstat;
int 		c, err, maxsize_set, maxlife_set;
int			do_rescan, do_expire, do_list, print_stat, do_update_param, print_books, is_profile, nfsen_format;
char		*maxsize_string, *lifetime_string, *datadir;
uint64_t	maxsize, lifetime, low_water;
channel_t	*channel, *current_channel;

	maxsize_string = lifetime_string = NULL;
	datadir = NULL;
	maxsize = lifetime = 0;
	do_rescan  		= 0;
	do_expire  		= 0;
	do_list	   		= 0;
	do_update_param = 0;
	is_profile		= 0;
	print_stat		= 0;
	print_books		= 0;
	maxsize_set 	= 0;
	maxlife_set 	= 0;
	low_water		= 0;
	nfsen_format	= 0;

	while ((c = getopt(argc, argv, "e:hl:L:Ypr:s:t:u:w:")) != EOF) {
		switch (c) {
			case 'h':
				usage(argv[0]);
				exit(0);
				break;
			case 'l':
				CheckDataDir(datadir);
				datadir = optarg;
				do_list = 1;
				print_stat = 1;
				break;
			case 'L':
				CheckDataDir(datadir);
				datadir = optarg;
				print_stat  = 1;
				print_books = 1;
				break;
			case 'p':
				is_profile = 1;
				break;
			case 'r':
				CheckDataDir(datadir);
				do_rescan = 1;
				print_stat = 1;
				datadir = optarg;
				break;
			case 'e':
				CheckDataDir(datadir);
				datadir = optarg;
				do_expire = 1;
				print_stat = 1;
				break;
			case 's':
				if ( ParseSizeDef(optarg, &maxsize ) == 0 )
					exit(250);
				maxsize_set = 1;
				break;
			case 't':
				if ( ParseTimeDef(optarg, &lifetime ) == 0 )
					exit(250);
				maxlife_set = 1;
				break;
			case 'u':
				CheckDataDir(datadir);
				datadir = optarg;
				do_update_param = 1;
				break;
			case 'w':
				low_water = strtoll(optarg, NULL, 10);
				if ( low_water > 100 ) {
					fprintf(stderr, "Water mark > 100%%\n");
					exit(250);
				}
				if ( low_water == 0 )
					low_water = 100;
				break;
			case 'Y':
				nfsen_format = 1;
				break;
			default:
				usage(argv[0]);
				exit(250);
		}

	}
	
	datadir = AbsolutePath(datadir);

	if ( !datadir ) {
		fprintf(stderr, "Missing data directory\n");
		usage(argv[0]);
		exit(250);
	}

	err  = stat(datadir, &fstat);
	if ( !(fstat.st_mode & S_IFDIR) ) {
		fprintf(stderr, "No such directory: %s\n", datadir);
		exit(250);
	}

	channel = GetChannelList(datadir, is_profile, do_rescan);
	// GetChannelList(datadir, is_profile, do_rescan);
	if ( !channel ) {
		exit(250);
	}

// printf("Size: %llu, time: %llu\n", maxsize, lifetime);

	// update water mark only, when not listing 
	if ( !is_profile && !do_list && low_water ) 
		channel->dirstat->low_water = low_water;

	/* process do_list first: if the UpdateStat results in a FORCE_REBUILD, 
	 * this will immediately done afterwards
	 * do_expire will need accurate books as well, so update the books here as well
	 */
	if ( do_list || do_expire ) {
		current_channel = channel;
		while ( current_channel ) {
			if ( current_channel->books_stat == BOOKKEEPER_OK ) {
				bookkeeper_t	tmp_books;
				printf("Include nfcapd bookeeping record in %s\n", current_channel->datadir);
				ClearBooks(current_channel->books, &tmp_books);
				UpdateStat(current_channel->dirstat, &tmp_books);
				if ( current_channel->dirstat->status == FORCE_REBUILD ) 
					current_channel->do_rescan = 1;
			}
			current_channel = current_channel->next;
		}
	}

	// process do_rescan: make sure stats are up to date, if required 
	current_channel = channel;
	while ( current_channel ) {
		if ( current_channel->do_rescan ) {
			int	 i;
			uint64_t last_sequence;
	
			/* detect new files: If nfcapd adds a new file while we are rescanning the directory
		 	* this results in inconsistent data for the rescan. Therefore check at the begin and end
		 	* of the rescan for the sequence number, which reflects the accesss/change to the bookkeeping record
		 	* It's assumed, that such an event does not occure more than once. However, loop 3 times max
		 	*/
			for ( i=0; i<3; i++ ) {
				last_sequence = BookSequence(current_channel->books);
				printf("Scanning files in %s .. ", current_channel->datadir);
				RescanDir(current_channel->datadir, current_channel->dirstat);
				if ( current_channel->dirstat->numfiles == 0 ) { //nothing found
					current_channel->status = NOFILES;
				}
				if ( BookSequence(current_channel->books) == last_sequence ) 
					break;
				printf("Rescan again, due to file changes in directory!\n");
			}
			if ( BookSequence(current_channel->books) != last_sequence ) {
				fprintf(stderr, "Could not savely rescan the directory. Data is not consistent.\n");
				ReleaseBookkeeper(current_channel->books, DETACH_ONLY);
				if ( current_channel->status == OK )
					WriteStatInfo(current_channel->dirstat);
				exit(250);
			}
			printf("done.\n");
			if ( current_channel->books_stat == BOOKKEEPER_OK ) {
				printf("Updating nfcapd bookeeping records\n");
				ClearBooks(channel->books, NULL);
			}
		}
		current_channel = current_channel->next;
	}

	// now process do_expire if required
	if ( do_expire ) {
		dirstat_t	old_stat, current_stat;

		if ( is_profile ) {
			current_stat.status = 0;
			current_stat.max_lifetime 	= lifetime;
			current_stat.max_size  		= maxsize;
			current_stat.low_water 		= low_water ? low_water : 98;

			// sum up all channels in the profile
			current_channel = channel;
			current_stat.numfiles = current_channel->dirstat->numfiles;
			current_stat.filesize = current_channel->dirstat->filesize;
			current_stat.first 	  = current_channel->dirstat->first;
			current_stat.last 	  = current_channel->dirstat->last;
			current_channel = current_channel->next;
			while ( current_channel ) {
				current_stat.numfiles += current_channel->dirstat->numfiles;
				current_stat.filesize += current_channel->dirstat->filesize;
				if ( current_channel->dirstat->first && (current_channel->dirstat->first < current_stat.first) ) 
					current_stat.first = current_channel->dirstat->first;
				if ( current_channel->dirstat->last > current_stat.last ) 
					current_stat.last = current_channel->dirstat->last;
				
				current_channel = current_channel->next;
			}
			old_stat = current_stat;
			ExpireProfile(channel, &current_stat, maxsize, lifetime);

		} else {
			// cmd args override dirstat values
			if ( maxsize_set ) 
				channel->dirstat->max_size     = maxsize;
			else 
				maxsize = channel->dirstat->max_size;
			if ( maxlife_set ) 
				channel->dirstat->max_lifetime = lifetime;
			else 
				lifetime = channel->dirstat->max_lifetime;
	
		
			old_stat = *(channel->dirstat);
			ExpireDir(channel->datadir, channel->dirstat, maxsize, lifetime);
			current_stat = *(channel->dirstat);

		}
		// Report, what we have done
		printf("Expired files:      %llu\n", (unsigned long long)(old_stat.numfiles - current_stat.numfiles));
		printf("Expired file size:  %s\n", ScaleValue(old_stat.filesize - current_stat.filesize));
		printf("Expired time range: %s\n\n", ScaleTime(current_stat.first - old_stat.first));
	}

	if ( !is_profile && do_update_param ) {
		switch (channel->books_stat) {
			case BOOKKEEPER_OK:
				if ( maxsize_set ) 
					channel->dirstat->max_size = maxsize;
				else 
					maxsize = channel->dirstat->max_size;
				if ( maxlife_set ) 
					channel->dirstat->max_lifetime = lifetime;
				else 
					lifetime = channel->dirstat->max_lifetime;
				printf("Update collector process running for directory: '%s'\n", datadir);
				UpdateBooksParam(channel->books, (time_t)lifetime, maxsize);
				print_stat = 1;
				break;
			case ERR_NOTEXISTS:
				if ( maxsize_set ) 
					channel->dirstat->max_size = maxsize;
				if ( maxlife_set ) 
					channel->dirstat->max_lifetime = lifetime;
				print_stat = 1;
				break;
			default:
				// should never be reached as already cought earlier
				printf("Error %i while connecting to collector\n", channel->books_stat);
		}
	}

	if ( !is_profile && print_books ) {
		switch (channel->books_stat) {
			case BOOKKEEPER_OK:
				PrintBooks(channel->books);
				break;
			case ERR_NOTEXISTS:
				printf("No collector process running for directory: '%s'\n", channel->datadir);
				break;
			default:
				// should never be reached as already cought earlier
				printf("Error %i while connecting to collector\n", channel->books_stat);
		}

	}

	if ( print_stat ) {
		if ( is_profile ) {
			dirstat_t	current_stat;

			current_stat.status = 0;
			current_stat.max_lifetime 	= lifetime;
			current_stat.max_size  		= maxsize;
			current_stat.low_water 		= low_water ? low_water : 98;

			// sum up all channels in the profile
			current_channel = channel;
			current_stat.numfiles = current_channel->dirstat->numfiles;
			current_stat.filesize = current_channel->dirstat->filesize;
			current_stat.first 	  = current_channel->dirstat->first;
			current_stat.last 	  = current_channel->dirstat->last;
			current_channel = current_channel->next;
			while ( current_channel ) {
				current_stat.numfiles += current_channel->dirstat->numfiles;
				current_stat.filesize += current_channel->dirstat->filesize;
				if ( current_channel->dirstat->first && (current_channel->dirstat->first < current_stat.first) ) 
					current_stat.first = current_channel->dirstat->first;
				if ( current_channel->dirstat->last > current_stat.last ) 
					current_stat.last = current_channel->dirstat->last;
				
				current_channel = current_channel->next;
			}
			if ( nfsen_format ) {
				printf("Stat|%llu|%llu|%llu\n", 
					(unsigned long long)current_stat.filesize, 
					(unsigned long long)current_stat.first, (unsigned long long)current_stat.last);
			} else 
				PrintDirStat(&current_stat);
	} else
			if ( nfsen_format )
				printf("Stat|%llu|%llu|%llu\n", 
					(unsigned long long)channel->dirstat->filesize, 
					(unsigned long long)channel->dirstat->first, (unsigned long long)channel->dirstat->last );
			else 
				PrintDirStat(channel->dirstat);
			
	}

	current_channel = channel;
	while ( current_channel ) {
		ReleaseBookkeeper(current_channel->books, DETACH_ONLY);
		if ( current_channel->status == OK )
			WriteStatInfo(current_channel->dirstat);

		current_channel = current_channel->next;
	}

	return 0;
} 
